### ISSUE TITLE
Feature Request: Alternative `register_type` API allowing one level of recursion

### ISSUE THREAD

**Issue by casperisfine** [msgpack/msgpack-ruby#224]
Note: I'm saying feature request but I'm willing to implement it myself. I had a quick look and I think it's doable, but I'd rather get a bit of feedback first as it's non-trivial.

### Context

msgpack's extension types are great, but sometimes a bit limited. It work well for simple value types like `Time` etc, but in many cases serializing a complex object to a binary string is tricky.

What we've been doing with @shioyama is basically use msgpack in msgpack, eg.:

```ruby
MyFactory.register_type(
  0x12,
  SomeObject,
  packer: ->(obj) { MyFactory.dump(obj.to_h) },
  unpacker ->(payload) { SomeObject.load(MyFactory.load(payload)) }
)
```

The above is total pseudo code, but hopefully it help conveying the idea.

The problem with that is that it instantiate many new packers/unpackers for individual values, it is also a bit tricky to ensure we don't recurse.

### Feature Request

It would be great if msgpack had a similar feature built-in, as I think it could be implemented much more efficiently. e.g something like:

```ruby
MyFactory.register_meta_type(
  0x12,
  SomeObject,
  packer: &:as_msgpack,
  unpacker: ->(hash) SomeObject.load_from_msgpack(hash),
)
```

The `packer` would return simple Ruby types, and msgpack would take care of serializing them. Of course to avoid unbound recursion, it wouldn't be allowed to return "meta_types", only builtin and regular extension types.

@tagomoris is this a feature you'd be open to if I were to open a PR for it?

**Comment by tagomoris** [comment]
IIUC, this feature request is for a kind of container objects, like Hash and Array.
In the case of Hash and Array, msgpack-ruby calls `#to_msgpack` of contained objects recursively, and `#to_msgpack` receives `Packer` or `IO`(-like) objects as its argument. It can prevent instantiating packers recursively.
(See https://github.com/msgpack/msgpack-ruby/blob/master/lib/msgpack/core_ext.rb )

So, is it possible and useful for the case here?
If `packer` and `unpacker` procs can receive `self` object (`Packer` or `Unpacker` in those context) as `self` or the 2nd argument, we can use it in the recursive packing/unpacking without registering special types or introducing special mechanism for recursive packing/unpacking.

@casperisfine What do you think about this idea?

**Comment by casperisfine** [comment]
Implemented in https://github.com/msgpack/msgpack-ruby/pull/261, and released in 1.5.0 (1.5.1 has a fix for it though). Closing.
